modules/storage/storage-account/main.bicep

```diff
@@ -19,6 +19,142 @@ param isZoneRedundant bool = false
+param roleAssignments roleAssignmentsArray = []
+
+type blobServiceProperties = {
+  changeFeed: changeFeed?
+  containerDeleteRetentionPolicy: containerDeleteRetentionPolicy?
+  cors: cors?
+  deleteRetentionPolicy: deleteRetentionPolicy?
+  isVersioningEnabled: isBlobVersioningEnabled?
+  lastAccessTimeTrackingPolicy: lastAccessTimeTrackingPolicy?
+  restorePolicy: restorePolicy?
+}
+
+type changeFeed = {
+  enabled: bool
+  retentionInDays: changeFeedretentionInDays?
+}
+
+@minValue(1)
+@maxValue(146000)
+type changeFeedretentionInDays = int
+
+type containerDeleteRetentionPolicy = {
+  allowPermanentDelete: bool
+  days: containerDeleteRetensionPolicyDays?
+  enabled: bool
+}
+
+@minValue(1)
+@maxValue(365)
+type containerDeleteRetensionPolicyDays = int
+
+type cors = {
+  corsRules: [
+    {
+      allowedHeaders: [
+        'string'
+      ]
+      allowedMethods: [
+        'string'
+      ]
+      allowedOrigins: [
+        'string'
+      ]
+      exposedHeaders: [
+        'string'
+      ]
+      maxAgeInSeconds: int
+    }
+  ]
+}
+
+type deleteRetentionPolicy = {
+  allowPermanentDelete: bool
+  days: deleteRetentionPolicyDays?
+  enabled: bool
+}
+
+@minValue(1)
+@maxValue(365)
+type deleteRetentionPolicyDays = int
+
+@description('Toggle to enable or disable versioning for Blob service of the Storage Account. Used only if enableBlobService is set to true.')
+type isBlobVersioningEnabled = bool
+
+type lastAccessTimeTrackingPolicy = {
+  blobType: [
+    'string'
+  ]
+  enable: bool
+  name: 'AccessTimeTracking'
+  trackingGranularityInDays: trackingGranularityInDays?
+}
+
+type trackingGranularityInDays = 1
+
+type restorePolicy = {
+  days: restorePolicyDays?
+  enabled: bool
+}
+
+@minValue(1)
+@maxValue(365)
+type restorePolicyDays = int
+
+type blobServiceContainerProperties = {
+  defaultEncryptionScope: defaultEncryptionScope?
+  denyEncryptionScopeOverride: denyEncryptionScopeOverride?
+  immutableStorageWithVersioning: immutableStorageWithVersioning?
+  publicAccess: blobContainerPublicAccess?
+}
+
+@description('Allowed values are \'Blob\', \'Container\' or \'None\'')
+type blobContainerPublicAccess = string
+
+type defaultEncryptionScope = string
+
+type denyEncryptionScopeOverride = string
+
+type immutableStorageWithVersioning = {
+  enabled: bool
+}
+
+type roleAssignmentsArray = {
+  description: roleAssignmentDescription?
+  roleDefinitionIdOrName: roleDefinitionIdOrName?
+  principalIds: principalIds?
+  principalType: principalType?
+  resourceType: resourceType?
+}[]
+
+type roleAssignmentDescription = string
+
+type roleDefinitionIdOrName = string
+
+type principalIds = string[]
+
+type principalType = string
+
+type resourceType = string      
```

Comment:
```suggestion
type roleAssignmentsArray = {
  description: string?
  roleDefinitionIdOrName: string?
  principalIds: string[]?
  principalType: string?
  resourceType: string?
}[]
 
```

Reply:
we haven't come up with any style guidelines for custom typing yet... but I don't think there is an advantage to breaking it up further.

Reply:
@jeskew , could you give us some feedback on this PR, particularly any suggestions for how to write the custom types? Want to be consistent with people as we review the usage.

Reply:
changed my mind... let's complete this now and circle back around if we want to refactor types.